File: graph.py
"""Graph building and place mapping from OSM data."""

from collections import defaultdict
from multiprocessing import Pool, cpu_count

from rich.console import Console
from rich.progress import Progress, SpinnerColumn, BarColumn, TextColumn

from config import (
    Place,
    ROAD_SPEED,
    INTERSECTION_DELAY,
    POPULAR_AMENITY,
    POPULAR_SHOP,
    POPULAR_LEISURE,
    POPULAR_TOURISM,
)
from geo import haversine

console = Console()

# Module-level variables for worker processes
_graph_list = None
_node_coords = None


def parse_overpass_to_graph(data):
    """Convert Overpass JSON into nodes, edges (time-weighted), and named places."""
    if data is None or "elements" not in data:
        console.print("[red]No data from Overpass.[/red]")
        return {}, [], []

    elements = data["elements"]

    # Extract node coordinates and tags
    node_coords = {}
    for el in elements:
        if el.get("type") == "node":
            nid = str(el["id"])
            node_coords[nid] = (el.get("lat"), el.get("lon"))

    edges = []
    named_places_raw = []

    console.print("[yellow]Processing ways and named places...[/yellow]")

    for el in elements:
        if el.get("type") == "way":
            edges.extend(_process_way(el, node_coords))
        elif el.get("type") == "node":
            place = _extract_place(el, node_coords)
            if place:
                named_places_raw.append(place)

    console.print(
        f"[green]✔ Nodes: {len(node_coords)}, edges: {len(edges)}, "
        f"named places: {len(named_places_raw)}[/green]\n"
    )
    return node_coords, edges, named_places_raw


def _process_way(el, node_coords):
    """Process a single way element into edges."""
    tags = el.get("tags", {})
    highway = tags.get("highway")
    
    if not highway or highway not in ROAD_SPEED:
        return []

    oneway = tags.get("oneway", "no")
    speed = ROAD_SPEED.get(highway, 40)
    nds = [str(nid) for nid in el.get("nodes", [])]
    edges = []

    for i in range(len(nds) - 1):
        u, v = nds[i], nds[i + 1]
        if u not in node_coords or v not in node_coords:
            continue
            
        lat1, lon1 = node_coords[u]
        lat2, lon2 = node_coords[v]
        dist_km = haversine(lat1, lon1, lat2, lon2)
        time_min = dist_km / (speed / 60.0) + INTERSECTION_DELAY

        if oneway == "yes":
            edges.append((u, v, time_min))
        elif oneway == "-1":
            edges.append((v, u, time_min))
        else:
            edges.append((u, v, time_min))
            edges.append((v, u, time_min))

    return edges


def _extract_place(el, node_coords):
    """Extract place info from a node if it's a popular category."""
    nid = str(el["id"])
    tags = el.get("tags", {})
    name = tags.get("name")
    
    if not name:
        return None

    category = None
    if tags.get("amenity") in POPULAR_AMENITY:
        category = tags["amenity"]
    elif tags.get("shop") in POPULAR_SHOP:
        category = f"shop:{tags['shop']}"
    elif tags.get("leisure") in POPULAR_LEISURE:
        category = tags["leisure"]
    elif tags.get("tourism") in POPULAR_TOURISM:
        category = tags["tourism"]

    if category and nid in node_coords:
        lat, lon = node_coords[nid]
        return (nid, name, category, lat, lon)
    return None


def build_graph(edges):
    """Build adjacency list graph from edges."""
    graph = defaultdict(list)
    for u, v, t in edges:
        graph[u].append((v, t))
    return graph


def _init_worker(graph_list, node_coords):
    """Initialize worker process with shared data."""
    global _graph_list, _node_coords
    _graph_list = graph_list
    _node_coords = node_coords


def _find_nearest_node(place_data):
    """Worker function to find nearest graph node for a single place."""
    pid, name, cat, plat, plon = place_data
    best_node = None
    best_dist = float("inf")
    
    for nid in _graph_list:
        nlat, nlon = _node_coords[nid]
        d = haversine(plat, plon, nlat, nlon)
        if d < best_dist:
            best_dist = d
            best_node = nid
            
    return Place(pid, name, cat, plat, plon, best_node)


def map_places_to_nearest_nodes(places_raw, graph_nodes, node_coords):
    """Map places to nearest road nodes using parallel processing."""
    graph_list = list(graph_nodes)
    if not graph_list:
        console.print("[red]No road nodes in graph![/red]")
        return []

    num_places = len(places_raw)
    num_workers = min(cpu_count(), num_places, 8)

    console.print(
        f"[yellow]Mapping {num_places} places to nearest road nodes "
        f"using {num_workers} workers...[/yellow]"
    )

    with Progress(
        SpinnerColumn(),
        TextColumn("[progress.description]{task.description}"),
        BarColumn(),
        TextColumn("[progress.percentage]{task.percentage:>3.0f}%"),
        console=console,
    ) as progress:
        task = progress.add_task("Mapping places...", total=num_places)

        with Pool(
            processes=num_workers,
            initializer=_init_worker,
            initargs=(graph_list, node_coords),
        ) as pool:
            mapped = []
            chunksize = max(1, num_places // (num_workers * 4))
            for result in pool.imap(_find_nearest_node, places_raw, chunksize=chunksize):
                mapped.append(result)
                progress.update(task, advance=1)

    console.print(f"[green]✔ Places mapped: {len(mapped)}[/green]\n")
    return mapped

File: web/js/graph.js
// Graph building and place mapping from OSM data

import { haversine } from './geo.js';
import { 
    ROAD_SPEED, 
    POPULAR_AMENITY, 
    POPULAR_SHOP, 
    POPULAR_LEISURE, 
    POPULAR_TOURISM 
} from './config.js';

/**
 * Place object representing a named location
 * @typedef {Object} Place
 * @property {string} id - OSM node ID
 * @property {string} name - Place name
 * @property {string} category - Place category
 * @property {number} lat - Latitude
 * @property {number} lon - Longitude
 * @property {string | null} nearestNode - Nearest road node ID
 * @property {string} street - Street/location info
 */

/**
 * Parse Overpass JSON into nodes, edges, and named places
 * @param {object} data - Overpass JSON response
 * @param {function} onProgress - Progress callback
 * @returns {{ nodeCoords: Map, edges: Array, placesRaw: Array }}
 */
export function parseOverpassToGraph(data, onProgress = null) {
    if (!data || !data.elements) {
        return { nodeCoords: new Map(), edges: [], placesRaw: [] };
    }

    const elements = data.elements;
    
    // Extract node coordinates
    if (onProgress) onProgress('Extracting node coordinates...');
    const nodeCoords = new Map();
    for (const el of elements) {
        if (el.type === 'node') {
            nodeCoords.set(String(el.id), [el.lat, el.lon]);
        }
    }

    const edges = [];
    const placesRaw = [];

    if (onProgress) onProgress('Processing ways and places...');
    
    for (const el of elements) {
        if (el.type === 'way') {
            const wayEdges = processWay(el, nodeCoords);
            edges.push(...wayEdges);
        } else if (el.type === 'node') {
            const place = extractPlace(el, nodeCoords);
            if (place) {
                placesRaw.push(place);
            }
        }
    }

    return { nodeCoords, edges, placesRaw };
}

/**
 * Process a single way element into edges
 * @param {object} el - Way element
 * @param {Map} nodeCoords - Node coordinates map
 * @returns {Array} Array of edges [u, v, time]
 */
function processWay(el, nodeCoords) {
    const tags = el.tags || {};
    const highway = tags.highway;
    
    if (!highway || !(highway in ROAD_SPEED)) {
        return [];
    }

    const oneway = tags.oneway || 'no';
    const speed = ROAD_SPEED[highway] || 40;
    const nodes = (el.nodes || []).map(String);
    const edges = [];

    for (let i = 0; i < nodes.length - 1; i++) {
        const u = nodes[i];
        const v = nodes[i + 1];
        
        if (!nodeCoords.has(u) || !nodeCoords.has(v)) {
            continue;
        }

        const [lat1, lon1] = nodeCoords.get(u);
        const [lat2, lon2] = nodeCoords.get(v);
        const distKm = haversine(lat1, lon1, lat2, lon2);
        const timeMin = distKm / (speed / 60.0);

        if (oneway === 'yes') {
            edges.push([u, v, timeMin]);
        } else if (oneway === '-1') {
            edges.push([v, u, timeMin]);
        } else {
            edges.push([u, v, timeMin]);
            edges.push([v, u, timeMin]);
        }
    }

    return edges;
}

/**
 * Extract place info from a node if it's a popular category
 * @param {object} el - Node element
 * @param {Map} nodeCoords - Node coordinates map
 * @returns {object | null} Place info or null
 */
function extractPlace(el, nodeCoords) {
    const nid = String(el.id);
    const tags = el.tags || {};
    const name = tags.name;
    
    if (!name) {
        return null;
    }

    let category = null;
    if (tags.amenity && POPULAR_AMENITY.has(tags.amenity)) {
        category = tags.amenity;
    } else if (tags.shop && POPULAR_SHOP.has(tags.shop)) {
        category = `shop:${tags.shop}`;
    } else if (tags.leisure && POPULAR_LEISURE.has(tags.leisure)) {
        category = tags.leisure;
    } else if (tags.tourism && POPULAR_TOURISM.has(tags.tourism)) {
        category = tags.tourism;
    }

    if (category && nodeCoords.has(nid)) {
        const [lat, lon] = nodeCoords.get(nid);
        const location = 
            tags['addr:street'] ||
            tags['addr:district'] ||
            tags['addr:suburb'] ||
            tags['addr:neighbourhood'] ||
            tags['addr:neighborhood'] ||
            tags['addr:city'] ||
            '';
        
        return {
            id: nid,
            name,
            category,
            lat,
            lon,
            nearestNode: null,
            street: location
        };
    }
    
    return null;
}

/**
 * Build adjacency list graph from edges
 * @param {Array} edges - Array of edges [u, v, time]
 * @returns {Map} Adjacency list graph
 */
export function buildGraph(edges) {
    const graph = new Map();
    
    for (const [u, v, t] of edges) {
        if (!graph.has(u)) {
            graph.set(u, []);
        }
        graph.get(u).push([v, t]);
    }
    
    return graph;
}

/**
 * Map places to nearest road nodes
 * @param {Array} placesRaw - Raw places array
 * @param {Set} graphNodes - Set of graph node IDs
 * @param {Map} nodeCoords - Node coordinates map
 * @param {function} onProgress - Progress callback
 * @returns {Array} Array of Place objects with nearestNode set
 */
export function mapPlacesToNearestNodes(placesRaw, graphNodes, nodeCoords, onProgress = null) {
    const graphList = Array.from(graphNodes);
    
    if (graphList.length === 0) {
        return [];
    }

    const mapped = [];
    const total = placesRaw.length;

    for (let i = 0; i < placesRaw.length; i++) {
        const place = placesRaw[i];
        
        if (onProgress && i % 50 === 0) {
            onProgress(`Mapping places... ${Math.round((i / total) * 100)}%`);
        }

        let bestNode = null;
        let bestDist = Infinity;

        for (const nid of graphList) {
            const [nlat, nlon] = nodeCoords.get(nid);
            const d = haversine(place.lat, place.lon, nlat, nlon);
            if (d < bestDist) {
                bestDist = d;
                bestNode = nid;
            }
        }

        mapped.push({
            ...place,
            nearestNode: bestNode
        });
    }

    return mapped;
}

/**
 * Group places by category
 * @param {Array} places - Array of Place objects
 * @returns {Map} Map of category -> places array
 */
export function groupPlacesByCategory(places) {
    const categories = new Map();
    
    for (const place of places) {
        if (!categories.has(place.category)) {
            categories.set(place.category, []);
        }
        categories.get(place.category).push(place);
    }
    
    return categories;
}
